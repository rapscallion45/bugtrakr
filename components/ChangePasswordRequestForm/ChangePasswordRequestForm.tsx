import { FC } from 'react';
import Container from '@mui/material/Container';
import Card from '@mui/material/Card';
import Typography from '@mui/material/Typography';
import Button from '@mui/material/Button';
import CircularProgress from '@mui/material/CircularProgress';
import useChangePasswordRequestFormController from './ChangePasswordRequestFormController';

const ChangePasswordRequestForm: FC = function ChangePasswordRequestForm() {
  const { requestingPassword, demo, changePasswordRequest } =
    useChangePasswordRequestFormController();

  return (
    <Container maxWidth="sm" sx={{ marginTop: '30px' }}>
      <Card sx={{ display: 'flex', flexDirection: 'column', padding: '50px 20px' }}>
        <Typography variant="h5" component="h5" sx={{ textAlign: 'center' }}>
          Change Password Request
        </Typography>
        <Typography variant="body2" component="p" sx={{ padding: '20px 0', textAlign: 'center' }}>
          Clicking the link below will generate a new password reset link, which will be sent to
          your registered email address. Follow the link in the email to reset your password.
        </Typography>

        <Button
          type="button"
          fullWidth
          variant="contained"
          color="primary"
          disabled={requestingPassword || demo}
          onClick={changePasswordRequest}
          sx={{ margin: '20px 0', padding: '10px 0', alignSelf: 'center', maxWidth: '300px' }}
        >
          {!requestingPassword && 'Change Password'}
          {requestingPassword && <CircularProgress size={25} color="inherit" />}
        </Button>
      </Card>
    </Container>
  );
};

export default ChangePasswordRequestForm;
