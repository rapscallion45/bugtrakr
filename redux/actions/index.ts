export { default as alertActions } from './alert.actions';
export { default as accountActions } from './account.actions';
