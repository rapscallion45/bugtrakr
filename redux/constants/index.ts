export { default as alertConstants } from './alert.constants';
export { default as accountConstants } from './account.constants';
