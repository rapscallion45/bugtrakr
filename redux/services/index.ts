export { default as accountService } from './account.service';
export { default as projectsService } from './projects.service';
