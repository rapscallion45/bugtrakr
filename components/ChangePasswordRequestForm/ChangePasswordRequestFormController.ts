import { useSelector, useDispatch } from 'react-redux';
import { accountActions } from '../../redux/actions';

const useChangePasswordRequestFormController = () => {
  const dispatch = useDispatch();
  const { requestingPassword } = useSelector((state) => state.changePassword);
  const { user, demo } = useSelector((state) => state.authentication);
  const { email } = user;

  const changePasswordRequest = () => {
    dispatch(accountActions.changePassword(email));
  };

  return {
    requestingPassword,
    demo,
    changePasswordRequest,
  };
};
export default useChangePasswordRequestFormController;
