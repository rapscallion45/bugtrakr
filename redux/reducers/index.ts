import { combineReducers } from 'redux';

import registration from './registration.reducer';
import verifyEmail from './verifyEmail.reducer';
import changePassword from './changePassword.reducer';
import resetPassword from './resetPassword.reducer';
import validateResetToken from './validateResetToken.reducer';
import account from './account.reducer';
import alert from './alert.reducer';
import projects from './projects.reducer';
import bugs from './bugs.reducer';
import users from './users.reducer';

const rootReducer = combineReducers({
  account,
  registration,
  verifyEmail,
  changePassword,
  resetPassword,
  validateResetToken,
  alert,
  projects,
  bugs,
  users,
});
export default rootReducer;

export type AppState = ReturnType<typeof rootReducer>;
