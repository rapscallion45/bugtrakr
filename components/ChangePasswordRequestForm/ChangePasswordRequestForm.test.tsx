import React from 'react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { mount, render } from 'enzyme';
import toJson from 'enzyme-to-json';
import { waitFor } from '@testing-library/react';
import rootReducer from '../../redux/reducers';
import ChangePasswordRequestForm from './ChangePasswordRequestForm';
import accountMock from '../../__mocks__/accountMock';

const middleware = [thunkMiddleware];

describe('Change Password Request Form', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Renders correctly enzyme', () => {
    /* Arrange */
    const testStore = createStore(
      rootReducer,
      {
        authentication: {
          authenticating: false,
          loggedIn: true,
          user: {
            email: accountMock.email,
          },
        },
      },
      applyMiddleware(...middleware)
    );

    /* Act */
    const view = render(
      <Provider store={testStore}>
        <ChangePasswordRequestForm />
      </Provider>
    );

    /* Assert */
    expect(toJson(view)).toMatchSnapshot();
  });

  describe('Form Submission', () => {
    it('Should be disabled and spinner should be rendered when forgot password API called', () => {
      /* Arrange */
      const testStore = createStore(
        rootReducer,
        {
          authentication: {
            authenticating: false,
            loggedIn: true,
            demo: false,
            user: {
              email: accountMock.email,
            },
          },
          changePassword: {
            requestingPassword: true,
          },
        },
        applyMiddleware(...middleware)
      );

      /* Act */
      const view = render(
        <Provider store={testStore}>
          <ChangePasswordRequestForm />
        </Provider>
      );

      /* Assert */
      expect(view.find('button').prop('disabled')).toBe(true);
      expect(view.find('circle').length).toBe(1);
    });

    it('Should be enabled with no spinner once forgot password API call completed', async () => {
      /* Arrange */
      const testStore = createStore(
        rootReducer,
        {
          authentication: {
            authenticating: false,
            loggedIn: true,
            demo: false,
            user: {
              email: accountMock.email,
            },
          },
        },
        applyMiddleware(...middleware)
      );

      /* Act */
      const view = mount(
        <Provider store={testStore}>
          <ChangePasswordRequestForm />
        </Provider>
      );
      view.find('button').simulate('click');

      /* Assert */
      expect(view.find('button').prop('disabled')).toBe(true);
      expect(view.find('circle').length).toBe(1);

      await waitFor(() => {
        view.update();
        expect(view.find('button').prop('disabled')).toBe(false);
      });
      expect(view.find('circle').length).toBe(0);
    });

    it('Should be disabled, with no spinner, when in demo mode', () => {
      /* Arrange */
      const testStore = createStore(
        rootReducer,
        {
          authentication: {
            authenticating: false,
            loggedIn: true,
            demo: true,
            user: {
              email: accountMock.email,
            },
          },
          changePassword: {
            requestingPassword: false,
          },
        },
        applyMiddleware(...middleware)
      );

      /* Act */
      const view = render(
        <Provider store={testStore}>
          <ChangePasswordRequestForm />
        </Provider>
      );

      /* Assert */
      expect(view.find('button').prop('disabled')).toBe(true);
      expect(view.find('circle').length).toBe(0);
    });
  });

  describe('Events', () => {
    beforeAll(() => jest.spyOn(window, 'fetch'));

    it('Should call forgot password API on button click, passing logged in user email', async () => {
      /* Arrange */
      const testStore = createStore(
        rootReducer,
        {
          authentication: {
            authenticating: false,
            loggedIn: true,
            user: {
              email: accountMock.email,
            },
          },
        },
        applyMiddleware(...middleware)
      );
      const { email } = accountMock;

      /* Act */
      const view = mount(
        <Provider store={testStore}>
          <ChangePasswordRequestForm />
        </Provider>
      );
      view.find('button').simulate('click');

      /* Assert */
      expect(view.find('button').prop('disabled')).toBe(true);
      expect(view.find('circle').length).toBe(1);

      await waitFor(() => {
        expect(window.fetch).toHaveBeenCalledTimes(1);
      });
      expect(window.fetch).toHaveBeenCalledWith(
        '/api/forgot-password',
        expect.objectContaining({
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ email }),
        })
      );
    });
  });
});
