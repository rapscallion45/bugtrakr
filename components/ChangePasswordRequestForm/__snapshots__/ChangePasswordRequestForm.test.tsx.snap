// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Change Password Request Form Renders correctly enzyme 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthSm css-wpa43l-MuiContainer-root"
>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-rc1clv-MuiPaper-root-MuiCard-root"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 css-ph2xsj-MuiTypography-root"
    >
      Change Password Request
    </h5>
    <p
      class="MuiTypography-root MuiTypography-body2 css-16pdcnq-MuiTypography-root"
    >
      Clicking the link below will generate a new password reset link, which will be sent to your registered email address. Follow the link in the email to reset your password.
    </p>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-1hpg5mo-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Change Password
    </button>
  </div>
</div>
`;
